### PR TITLE
Small builder refactors and a Location data class

### PR DESCRIPTION
**Pull request by ItsIlya**
**The Refactors**
- ChestGui -> ContainerGui
ChestGuiContainer -> ContainerGuiScreen
Container sounds about right for that type of Container-based Gui instead of just calling it ChestGui

- SidebarInfo -> ScoreboardDisplay
Same thing here, ~~Sidebar~~ it's called Scoreboard! And it is definitely not going to be used only for showing info

**Location**
Location data class makes it easier to work with a location in a world.
It contains the position (Vec3d), a dimension (Identifier), a rotation (Vec2f) and a getter for getting the dimension as World.

### CODE REVIEW

File: gunpowder-api/src/main/kotlin/io/github/gunpowder/api/util/Location.kt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        val position: Vec3d,
+        val rotation: Vec2f,
+        val dimension: Identifier
+) {
+    companion object {
+        @JvmStatic
+        fun of(entity: Entity) = Location(entity.pos, Vec2f(entity.yaw, entity.pitch), entity.world.registryKey.value)
+    }
+
+    val world: ServerWorld?
```
Comment (Martmists-GH) [inline gunpowder-api/src/main/kotlin/io/github/gunpowder/api/util/Location.kt]: Move properties to before companion objects

File: gunpowder-base/src/main/kotlin/io/github/gunpowder/entities/GunpowderRegistry.kt [hunk 2]
```diff
@@ -39,12 +39,11 @@ import net.minecraft.server.command.ServerCommandSource
 import org.jetbrains.exposed.sql.SchemaUtils
 import org.jetbrains.exposed.sql.Table
 import org.jetbrains.exposed.sql.transactions.transaction
-import java.io.File
 import java.util.function.Supplier
 import io.github.gunpowder.api.GunpowderRegistry as APIGunpowderRegistry
-import io.github.gunpowder.api.builders.ChestGui as APIChestGui
+import io.github.gunpowder.api.builders.ContainerGui as APIChestGui
```
Comment (Martmists-GH) [inline gunpowder-base/src/main/kotlin/io/github/gunpowder/entities/GunpowderRegistry.kt]: Update imported names too

File: gunpowder-base/src/main/kotlin/io/github/gunpowder/entities/GunpowderRegistry.kt [hunk 3]
```diff
@@ -39,12 +39,11 @@ import net.minecraft.server.command.ServerCommandSource
...
 import io.github.gunpowder.api.builders.Command as APICommand
-import io.github.gunpowder.api.builders.SidebarInfo as APISidebarInfo
+import io.github.gunpowder.api.builders.ScoreboardDisplay as APISidebarInfo
```
Comment (Martmists-GH) [inline gunpowder-base/src/main/kotlin/io/github/gunpowder/entities/GunpowderRegistry.kt]: Ditto

File: gunpowder-api/src/main/kotlin/io/github/gunpowder/api/util/Location.kt [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import io.github.gunpowder.api.GunpowderMod
+import net.minecraft.entity.Entity
+import net.minecraft.server.world.ServerWorld
+import net.minecraft.util.Identifier
+import net.minecraft.util.math.Vec2f
+import net.minecraft.util.math.Vec3d
+import net.minecraft.util.registry.Registry
+import net.minecraft.util.registry.RegistryKey
+
+data class Location(
```
Comment (Martmists-GH) [inline gunpowder-api/src/main/kotlin/io/github/gunpowder/api/util/Location.kt:43]: Try to implement uses for this in various utility functions or extension methods

**Review by Martmists-GH — CHANGES_REQUESTED** [review]

File: gunpowder-api/src/main/kotlin/io/github/gunpowder/api/util/Location.kt [hunk 5]
```diff
@@ -32,18 +32,22 @@ import net.minecraft.util.math.Vec2f
 import net.minecraft.util.math.Vec3d
 import net.minecraft.util.registry.Registry
 import net.minecraft.util.registry.RegistryKey
+import net.minecraft.world.World
 
 data class Location(
-        val position: Vec3d,
-        val rotation: Vec2f,
-        val dimension: Identifier
+        var position: Vec3d,
```
Comment (Martmists-GH) [inline gunpowder-api/src/main/kotlin/io/github/gunpowder/api/util/Location.kt]: These should not be mutable. Make functions to offset or change them into new objects.

File: gunpowder-api/src/main/kotlin/io/github/gunpowder/api/util/Location.kt [hunk 6]
```diff
@@ -32,18 +32,22 @@ import net.minecraft.util.math.Vec2f
...
+
     companion object {
         @JvmStatic
         fun of(entity: Entity) = Location(entity.pos, Vec2f(entity.yaw, entity.pitch), entity.world.registryKey.value)
-    }
 
-    val world: ServerWorld?
-        get() = GunpowderMod.instance.server.getWorld(RegistryKey.of(Registry.DIMENSION, dimension))
+        @JvmStatic
+        fun dummy() = Location(Vec3d.ZERO, Vec2f.ZERO, World.OVERWORLD.value)
```
Comment (Martmists-GH) [inline gunpowder-api/src/main/kotlin/io/github/gunpowder/api/util/Location.kt]: Make this a val instead.

**Review by Martmists-GH — CHANGES_REQUESTED** [review]

File: gunpowder-api/src/main/kotlin/io/github/gunpowder/api/util/Location.kt [hunk 7]
```diff
@@ -27,27 +27,84 @@ package io.github.gunpowder.api.util
...
+        val rotation: Vec2f?,
+        val dimension: Identifier
 ) {
     val world: ServerWorld?
         get() = GunpowderMod.instance.server.getWorld(RegistryKey.of(Registry.DIMENSION, dimension))
 
     companion object {
         @JvmStatic
-        fun of(entity: Entity) = Location(entity.pos, Vec2f(entity.yaw, entity.pitch), entity.world.registryKey.value)
+        val origin = Location(Vec3d.ZERO, Vec2f.ZERO, World.OVERWORLD.value)
```
Comment (Martmists-GH) [inline gunpowder-api/src/main/kotlin/io/github/gunpowder/api/util/Location.kt]: Not sure about this name, would like some more opinions on this. name should probably be capitalized.

File: gunpowder-api/src/main/kotlin/io/github/gunpowder/api/util/Location.kt [hunk 8]
```diff
@@ -27,27 +27,84 @@ package io.github.gunpowder.api.util
...
+    fun south() = offset(Direction.SOUTH)
+    fun south(distance: Int) = offset(distance, Direction.SOUTH)
+
+    fun west() = offset(Direction.WEST)
+    fun west(distance: Int) = offset(distance, Direction.WEST)
+
+    fun east() = offset(Direction.EAST)
+    fun east(distance: Int) = offset(distance, Direction.EAST)
+
+    fun offset(direction: Direction): Location = Location(
```
Comment (Martmists-GH) [inline gunpowder-api/src/main/kotlin/io/github/gunpowder/api/util/Location.kt]: fun offset(direction: Direction) = offset(1.0, direction)

File: gunpowder-api/src/main/kotlin/io/github/gunpowder/api/util/Location.kt [hunk 9]
```diff
@@ -27,27 +27,84 @@ package io.github.gunpowder.api.util
...
+            Vec3d(
+                    position.x + direction.offsetX, position.y + direction.offsetY, position.z + direction.offsetZ
+            ), rotation, dimension
+    )
+
+    fun offset(amount: Int, direction: Direction): Location = if (amount == 0) this else Location(
```
Comment (Martmists-GH) [inline gunpowder-api/src/main/kotlin/io/github/gunpowder/api/util/Location.kt]: No need for `: Location` type specifier

File: gunpowder-api/src/main/kotlin/io/github/gunpowder/api/util/Location.kt [hunk 10]
```diff
@@ -27,27 +27,84 @@ package io.github.gunpowder.api.util
...
+            Vec3d(
+                    position.x + direction.offsetX * amount,
+                    position.y + direction.offsetY * amount,
+                    position.z + direction.offsetZ * amount
+            ), rotation, dimension
+    )
+
+    fun axisAlignedRotate(rot: BlockRotation): Location {
```
Comment (Martmists-GH) [inline gunpowder-api/src/main/kotlin/io/github/gunpowder/api/util/Location.kt]: Again, no return type needed

File: gunpowder-api/src/main/kotlin/io/github/gunpowder/api/util/Location.kt [hunk 11]
```diff
@@ -27,27 +27,84 @@ package io.github.gunpowder.api.util
...
+    }
+
+    fun withRotation(yaw: Float, pitch: Float) = withRotation(Vec2f(yaw, pitch))
+    fun withRotation(vector: Vec2f) = Location(position, vector, dimension)
+
+    fun withDimension(world: World) = withDimension(world.registryKey.value)
+    fun withDimension(dimension: Identifier) = Location(position, rotation, dimension)
+
+    fun withPosition(x: Double, y: Double, z: Double) = withPosition(Vec3d(x, y, z))
+    fun withPosition(position: Vec3d) = Location(position, rotation, dimension)
```
Comment (Martmists-GH) [inline gunpowder-api/src/main/kotlin/io/github/gunpowder/api/util/Location.kt]: Add a vec3i argument option for Vec3i/BlockPos

File: gunpowder-base/src/main/kotlin/io/github/gunpowder/entities/builders/TeleportRequest.kt [hunk 12]
```diff
@@ -93,19 +93,19 @@ class TeleportRequest private constructor(
         }
 
         override fun facing(facing: Vec2f) {
-            this.destination.rotation = facing
+            this.destination.withRotation(facing)
```
Comment (Martmists-GH) [inline gunpowder-base/src/main/kotlin/io/github/gunpowder/entities/builders/TeleportRequest.kt]: You're not setting the values again, fix and test your code.

File: gunpowder-base/src/main/kotlin/io/github/gunpowder/entities/builders/TeleportRequest.kt [hunk 13]
```diff
@@ -93,19 +93,19 @@ class TeleportRequest private constructor(
...
         }
 
         override fun dimension(dimension: Identifier) {
-            this.destination.dimension = dimension
+            this.destination.withDimension(dimension)
```
Comment (Martmists-GH) [inline gunpowder-base/src/main/kotlin/io/github/gunpowder/entities/builders/TeleportRequest.kt]: Ditto

File: gunpowder-base/src/main/kotlin/io/github/gunpowder/entities/builders/TeleportRequest.kt [hunk 14]
```diff
@@ -93,19 +93,19 @@ class TeleportRequest private constructor(
...
         }
 
         override fun destination(destination: Location) {
             this.destination = destination
         }
 
         override fun destination(destination: Vec3d) {
-            this.destination.position = destination
+            this.destination
```
Comment (Martmists-GH) [inline gunpowder-base/src/main/kotlin/io/github/gunpowder/entities/builders/TeleportRequest.kt]: This method does nothing.

**Review by Martmists-GH — CHANGES_REQUESTED** [review]

**Comment by ItsIlya** [comment]
Fixed all those and tested.

**Comment by ItsIlya** [comment]
~~Javadoc~~ KDoc I meant o.o

File: gunpowder-api/src/main/kotlin/io/github/gunpowder/api/builders/TeleportRequest.kt [hunk 15]
```diff
@@ -38,9 +39,7 @@ import java.util.function.Consumer
 
 interface TeleportRequest {
     val player: ServerPlayerEntity
-    val destination: Vec3d
-    val dimension: Identifier
-    val facing: Vec2f?
+    val destination: Location
```
Comment (Martmists-GH) [inline gunpowder-api/src/main/kotlin/io/github/gunpowder/api/builders/TeleportRequest.kt:42]: Make sure to open a PR for this on Gunpowder Teleport as this is a breaking API change.

**Comment by Martmists-GH** [comment]
Needs reviewing again.

**Comment by Martmists-GH** [comment]
Closing for inactivity

**Pull request closed without merging**